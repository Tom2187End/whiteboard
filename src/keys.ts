export const isDarwin = /Mac|iPod|iPhone|iPad/.test(window.navigator.platform);

export const KEYS = {
  ARROW_LEFT: "ArrowLeft",
  ARROW_RIGHT: "ArrowRight",
  ARROW_DOWN: "ArrowDown",
  ARROW_UP: "ArrowUp",
  ENTER: "Enter",
  ESCAPE: "Escape",
  DELETE: "Delete",
  BACKSPACE: "Backspace",
  CTRL_OR_CMD: isDarwin ? "metaKey" : "ctrlKey",
  TAB: "Tab",
  SPACE: " ",
  QUESTION_MARK: "?",
  F_KEY_CODE: 70,
  ALT_KEY_CODE: 18,
  Z_KEY_CODE: 90,
  GRID_KEY_CODE: 222,
  G_KEY_CODE: 71,
  C_KEY_CODE: 67,
  V_KEY_CODE: 86,
} as const;

export type Key = keyof typeof KEYS;

export const isArrowKey = (keyCode: string) =>
  keyCode === KEYS.ARROW_LEFT ||
  keyCode === KEYS.ARROW_RIGHT ||
  keyCode === KEYS.ARROW_DOWN ||
  keyCode === KEYS.ARROW_UP;

export const getResizeCenterPointKey = (event: MouseEvent | KeyboardEvent) =>
  event.altKey || event.which === KEYS.ALT_KEY_CODE;

export const getResizeWithSidesSameLengthKey = (event: MouseEvent) =>
  event.shiftKey;

export const getRotateWithDiscreteAngleKey = (event: MouseEvent) =>
  event.shiftKey;
