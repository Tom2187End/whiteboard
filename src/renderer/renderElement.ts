import { ExcalidrawElement } from "../element/types";
import { isTextElement } from "../element/typeChecks";
import { getDiamondPoints, getArrowPoints } from "../element/bounds";
import { RoughCanvas } from "roughjs/bin/canvas";
import { Drawable } from "roughjs/bin/core";

export function renderElement(
  element: ExcalidrawElement,
  rc: RoughCanvas,
  context: CanvasRenderingContext2D
) {
  const generator = rc.generator;
  if (element.type === "selection") {
    const fillStyle = context.fillStyle;
    context.fillStyle = "rgba(0, 0, 255, 0.10)";
    context.fillRect(0, 0, element.width, element.height);
    context.fillStyle = fillStyle;
  } else if (element.type === "rectangle") {
    if (!element.shape) {
      element.shape = generator.rectangle(0, 0, element.width, element.height, {
        stroke: element.strokeColor,
        fill:
          element.backgroundColor === "transparent"
            ? undefined
            : element.backgroundColor,
        fillStyle: element.fillStyle,
        strokeWidth: element.strokeWidth,
        roughness: element.roughness,
        seed: element.seed
      });
    }

    context.globalAlpha = element.opacity / 100;
    rc.draw(element.shape as Drawable);
    context.globalAlpha = 1;
  } else if (element.type === "diamond") {
    if (!element.shape) {
      const [
        topX,
        topY,
        rightX,
        rightY,
        bottomX,
        bottomY,
        leftX,
        leftY
      ] = getDiamondPoints(element);
      element.shape = generator.polygon(
        [
          [topX, topY],
          [rightX, rightY],
          [bottomX, bottomY],
          [leftX, leftY]
        ],
        {
          stroke: element.strokeColor,
          fill:
            element.backgroundColor === "transparent"
              ? undefined
              : element.backgroundColor,
          fillStyle: element.fillStyle,
          strokeWidth: element.strokeWidth,
          roughness: element.roughness,
          seed: element.seed
        }
      );
    }

    context.globalAlpha = element.opacity / 100;
    rc.draw(element.shape as Drawable);
    context.globalAlpha = 1;
  } else if (element.type === "ellipse") {
    if (!element.shape) {
      element.shape = generator.ellipse(
        element.width / 2,
        element.height / 2,
        element.width,
        element.height,
        {
          stroke: element.strokeColor,
          fill:
            element.backgroundColor === "transparent"
              ? undefined
              : element.backgroundColor,
          fillStyle: element.fillStyle,
          strokeWidth: element.strokeWidth,
          roughness: element.roughness,
          seed: element.seed
        }
      );
    }

    context.globalAlpha = element.opacity / 100;
    rc.draw(element.shape as Drawable);
    context.globalAlpha = 1;
  } else if (element.type === "arrow") {
    const [x1, y1, x2, y2, x3, y3, x4, y4] = getArrowPoints(element);
    const options = {
      stroke: element.strokeColor,
      strokeWidth: element.strokeWidth,
      roughness: element.roughness,
      seed: element.seed
    };

    if (!element.shape) {
      element.shape = [
        //    \
        generator.line(x3, y3, x2, y2, options),
        // -----
        generator.line(x1, y1, x2, y2, options),
        //    /
        generator.line(x4, y4, x2, y2, options)
      ];
    }

    context.globalAlpha = element.opacity / 100;
    (element.shape as Drawable[]).forEach(shape => rc.draw(shape));
    context.globalAlpha = 1;
    return;
  } else if (isTextElement(element)) {
    context.globalAlpha = element.opacity / 100;
    const font = context.font;
    context.font = element.font;
    const fillStyle = context.fillStyle;
    context.fillStyle = element.strokeColor;
    context.fillText(
      element.text,
      0,
      element.baseline || element.actualBoundingBoxAscent || 0
    );
    context.fillStyle = fillStyle;
    context.font = font;
    context.globalAlpha = 1;
  } else {
    throw new Error("Unimplemented type " + element.type);
  }
}
