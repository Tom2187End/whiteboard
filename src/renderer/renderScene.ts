import { RoughCanvas } from "roughjs/bin/canvas";
import { RoughSVG } from "roughjs/bin/svg";

import { FlooredNumber, AppState } from "../types";
import { ExcalidrawElement } from "../element/types";
import {
  getElementAbsoluteCoords,
  OMIT_SIDES_FOR_MULTIPLE_ELEMENTS,
  handlerRectanglesFromCoords,
  handlerRectangles,
  getCommonBounds,
  canResizeMutlipleElements,
} from "../element";

import { roundRect } from "./roundRect";
import { SceneState } from "../scene/types";
import {
  getScrollBars,
  SCROLLBAR_COLOR,
  SCROLLBAR_WIDTH,
} from "../scene/scrollbars";
import { getSelectedElements } from "../scene/selection";

import { renderElement, renderElementToSvg } from "./renderElement";
import colors from "../colors";

type HandlerRectanglesRet = keyof ReturnType<typeof handlerRectangles>;

function colorsForClientId(clientId: string) {
  // Naive way of getting an integer out of the clientId
  const sum = clientId.split("").reduce((a, str) => a + str.charCodeAt(0), 0);

  // Skip transparent background.
  const backgrounds = colors.elementBackground.slice(1);
  const strokes = colors.elementStroke.slice(1);
  return {
    background: backgrounds[sum % backgrounds.length],
    stroke: strokes[sum % strokes.length],
  };
}

function strokeRectWithRotation(
  context: CanvasRenderingContext2D,
  x: number,
  y: number,
  width: number,
  height: number,
  cx: number,
  cy: number,
  angle: number,
  fill?: boolean,
) {
  context.translate(cx, cy);
  context.rotate(angle);
  if (fill) {
    context.fillRect(x - cx, y - cy, width, height);
  }
  context.strokeRect(x - cx, y - cy, width, height);
  context.rotate(-angle);
  context.translate(-cx, -cy);
}

function strokeCircle(
  context: CanvasRenderingContext2D,
  x: number,
  y: number,
  width: number,
  height: number,
) {
  context.beginPath();
  context.arc(x + width / 2, y + height / 2, width / 2, 0, Math.PI * 2);
  context.fill();
  context.stroke();
}

export function renderScene(
  allElements: readonly ExcalidrawElement[],
  appState: AppState,
  selectionElement: ExcalidrawElement | null,
  scale: number,
  rc: RoughCanvas,
  canvas: HTMLCanvasElement,
  sceneState: SceneState,
  // extra options, currently passed by export helper
  {
    renderScrollbars = true,
    renderSelection = true,
    // Whether to employ render optimizations to improve performance.
    // Should not be turned on for export operations and similar, because it
    //  doesn't guarantee pixel-perfect output.
    renderOptimizations = false,
  }: {
    renderScrollbars?: boolean;
    renderSelection?: boolean;
    renderOptimizations?: boolean;
  } = {},
) {
  if (!canvas) {
    return { atLeastOneVisibleElement: false };
  }

  const elements = allElements.filter((element) => !element.isDeleted);

  const context = canvas.getContext("2d")!;
  context.scale(scale, scale);

  // When doing calculations based on canvas width we should used normalized one
  const normalizedCanvasWidth = canvas.width / scale;
  const normalizedCanvasHeight = canvas.height / scale;

  // Paint background
  if (typeof sceneState.viewBackgroundColor === "string") {
    const hasTransparence =
      sceneState.viewBackgroundColor === "transparent" ||
      sceneState.viewBackgroundColor.length === 5 ||
      sceneState.viewBackgroundColor.length === 9;
    if (hasTransparence) {
      context.clearRect(0, 0, normalizedCanvasWidth, normalizedCanvasHeight);
    }
    const fillStyle = context.fillStyle;
    context.fillStyle = sceneState.viewBackgroundColor;
    context.fillRect(0, 0, normalizedCanvasWidth, normalizedCanvasHeight);
    context.fillStyle = fillStyle;
  } else {
    context.clearRect(0, 0, normalizedCanvasWidth, normalizedCanvasHeight);
  }

  // Apply zoom
  const zoomTranslationX = (-normalizedCanvasWidth * (sceneState.zoom - 1)) / 2;
  const zoomTranslationY =
    (-normalizedCanvasHeight * (sceneState.zoom - 1)) / 2;
  context.translate(zoomTranslationX, zoomTranslationY);
  context.scale(sceneState.zoom, sceneState.zoom);

  // Paint visible elements
  const visibleElements = elements.filter((element) =>
    isVisibleElement(
      element,
      normalizedCanvasWidth,
      normalizedCanvasHeight,
      sceneState,
    ),
  );

  visibleElements.forEach((element) => {
    renderElement(element, rc, context, renderOptimizations, sceneState);
  });

  // Pain selection element
  if (selectionElement) {
    renderElement(
      selectionElement,
      rc,
      context,
      renderOptimizations,
      sceneState,
    );
  }

  // Paint selected elements
  if (renderSelection) {
    context.translate(sceneState.scrollX, sceneState.scrollY);

    const selections = elements.reduce((acc, element) => {
      const selectionColors = [];
      // local user
      if (appState.selectedElementIds[element.id]) {
        selectionColors.push("#000000");
      }
      // remote users
      if (sceneState.remoteSelectedElementIds[element.id]) {
        selectionColors.push(
          ...sceneState.remoteSelectedElementIds[element.id].map((socketId) => {
            const { background } = colorsForClientId(socketId);
            return background;
          }),
        );
      }
      if (selectionColors.length) {
        acc.push({ element, selectionColors });
      }
      return acc;
    }, [] as { element: ExcalidrawElement; selectionColors: string[] }[]);

    selections.forEach(({ element, selectionColors }) => {
      const [
        elementX1,
        elementY1,
        elementX2,
        elementY2,
      ] = getElementAbsoluteCoords(element);

      const elementWidth = elementX2 - elementX1;
      const elementHeight = elementY2 - elementY1;

      const initialLineDash = context.getLineDash();
      const lineWidth = context.lineWidth;
      const lineDashOffset = context.lineDashOffset;
      const strokeStyle = context.strokeStyle;

      const dashedLinePadding = 4 / sceneState.zoom;
      const dashWidth = 8 / sceneState.zoom;
      const spaceWidth = 4 / sceneState.zoom;

      context.lineWidth = 1 / sceneState.zoom;

      const count = selectionColors.length;
      for (var i = 0; i < count; ++i) {
        context.strokeStyle = selectionColors[i];
        context.setLineDash([
          dashWidth,
          spaceWidth + (dashWidth + spaceWidth) * (count - 1),
        ]);
        context.lineDashOffset = (dashWidth + spaceWidth) * i;
        strokeRectWithRotation(
          context,
          elementX1 - dashedLinePadding,
          elementY1 - dashedLinePadding,
          elementWidth + dashedLinePadding * 2,
          elementHeight + dashedLinePadding * 2,
          elementX1 + elementWidth / 2,
          elementY1 + elementHeight / 2,
          element.angle,
        );
      }
      context.lineDashOffset = lineDashOffset;
      context.strokeStyle = strokeStyle;
      context.lineWidth = lineWidth;
      context.setLineDash(initialLineDash);
    });
    context.translate(-sceneState.scrollX, -sceneState.scrollY);

    const locallySelectedElements = getSelectedElements(elements, appState);

    // Paint resize handlers
    if (locallySelectedElements.length === 1) {
      context.translate(sceneState.scrollX, sceneState.scrollY);
      context.fillStyle = "#fff";
      const handlers = handlerRectangles(
        locallySelectedElements[0],
        sceneState.zoom,
      );
      Object.keys(handlers).forEach((key) => {
        const handler = handlers[key as HandlerRectanglesRet];
        if (handler !== undefined) {
          const lineWidth = context.lineWidth;
          context.lineWidth = 1 / sceneState.zoom;
          if (key === "rotation") {
            strokeCircle(
              context,
              handler[0],
              handler[1],
              handler[2],
              handler[3],
            );
          } else if (locallySelectedElements[0].type !== "text") {
            strokeRectWithRotation(
              context,
              handler[0],
              handler[1],
              handler[2],
              handler[3],
              handler[0] + handler[2] / 2,
              handler[1] + handler[3] / 2,
              locallySelectedElements[0].angle,
              true, // fill before stroke
            );
          }
          context.lineWidth = lineWidth;
        }
      });
      context.translate(-sceneState.scrollX, -sceneState.scrollY);
    } else if (locallySelectedElements.length > 1) {
      if (canResizeMutlipleElements(locallySelectedElements)) {
        const dashedLinePadding = 4 / sceneState.zoom;
        context.translate(sceneState.scrollX, sceneState.scrollY);
        context.fillStyle = "#fff";
        const [x1, y1, x2, y2] = getCommonBounds(locallySelectedElements);
        const initialLineDash = context.getLineDash();
        context.setLineDash([2 / sceneState.zoom]);
        const lineWidth = context.lineWidth;
        context.lineWidth = 1 / sceneState.zoom;
        strokeRectWithRotation(
          context,
          x1 - dashedLinePadding,
          y1 - dashedLinePadding,
          x2 - x1 + dashedLinePadding * 2,
          y2 - y1 + dashedLinePadding * 2,
          (x1 + x2) / 2,
          (y1 + y2) / 2,
          0,
        );
        context.lineWidth = lineWidth;
        context.setLineDash(initialLineDash);
        const handlers = handlerRectanglesFromCoords(
          [x1, y1, x2, y2],
          0,
          sceneState.zoom,
          undefined,
          OMIT_SIDES_FOR_MULTIPLE_ELEMENTS,
        );
        Object.keys(handlers).forEach((key) => {
          const handler = handlers[key as HandlerRectanglesRet];
          if (handler !== undefined) {
            const lineWidth = context.lineWidth;
            context.lineWidth = 1 / sceneState.zoom;
            strokeRectWithRotation(
              context,
              handler[0],
              handler[1],
              handler[2],
              handler[3],
              handler[0] + handler[2] / 2,
              handler[1] + handler[3] / 2,
              0,
              true, // fill before stroke
            );
            context.lineWidth = lineWidth;
          }
        });
        context.translate(-sceneState.scrollX, -sceneState.scrollY);
      }
    }
  }

  // Reset zoom
  context.scale(1 / sceneState.zoom, 1 / sceneState.zoom);
  context.translate(-zoomTranslationX, -zoomTranslationY);

  // Paint remote pointers
  for (const clientId in sceneState.remotePointerViewportCoords) {
    let { x, y } = sceneState.remotePointerViewportCoords[clientId];
    const username = sceneState.remotePointerUsernames[clientId];

    const width = 9;
    const height = 14;

    const isOutOfBounds =
      x < 0 ||
      x > normalizedCanvasWidth - width ||
      y < 0 ||
      y > normalizedCanvasHeight - height;

    x = Math.max(x, 0);
    x = Math.min(x, normalizedCanvasWidth - width);
    y = Math.max(y, 0);
    y = Math.min(y, normalizedCanvasHeight - height);

    const { background, stroke } = colorsForClientId(clientId);

    const strokeStyle = context.strokeStyle;
    const fillStyle = context.fillStyle;
    const globalAlpha = context.globalAlpha;
    context.strokeStyle = stroke;
    context.fillStyle = background;
    if (isOutOfBounds) {
      context.globalAlpha = 0.2;
    }

    if (
      sceneState.remotePointerButton &&
      sceneState.remotePointerButton[clientId] === "down"
    ) {
      context.beginPath();
      context.arc(x, y, 15, 0, 2 * Math.PI, false);
      context.lineWidth = 3;
      context.strokeStyle = "#ffffff88";
      context.stroke();
      context.closePath();

      context.beginPath();
      context.arc(x, y, 15, 0, 2 * Math.PI, false);
      context.lineWidth = 1;
      context.strokeStyle = stroke;
      context.stroke();
      context.closePath();
    }

    context.beginPath();
    context.moveTo(x, y);
    context.lineTo(x + 1, y + 14);
    context.lineTo(x + 4, y + 9);
    context.lineTo(x + 9, y + 10);
    context.lineTo(x, y);
    context.fill();
    context.stroke();

    if (!isOutOfBounds && username) {
      const offsetX = x + width;
      const offsetY = y + height;
      const paddingHorizontal = 4;
      const paddingVertical = 4;
      const measure = context.measureText(username);
      const measureHeight =
        measure.actualBoundingBoxDescent + measure.actualBoundingBoxAscent;

      // Border
      context.fillStyle = stroke;
      context.globalAlpha = globalAlpha;
      context.fillRect(
        offsetX - 1,
        offsetY - 1,
        measure.width + 2 * paddingHorizontal + 2,
        measureHeight + 2 * paddingVertical + 2,
      );
      // Background
      context.fillStyle = background;
      context.fillRect(
        offsetX,
        offsetY,
        measure.width + 2 * paddingHorizontal,
        measureHeight + 2 * paddingVertical,
      );
      context.fillStyle = "#ffffff";
      context.fillText(
        username,
        offsetX + paddingHorizontal,
        offsetY + paddingVertical + measure.actualBoundingBoxAscent,
      );
    }

    context.strokeStyle = strokeStyle;
    context.fillStyle = fillStyle;
    context.globalAlpha = globalAlpha;
    context.closePath();
  }

  // Paint scrollbars
  let scrollBars;
  if (renderScrollbars) {
    scrollBars = getScrollBars(
      elements,
      normalizedCanvasWidth,
      normalizedCanvasHeight,
      sceneState,
    );

    const fillStyle = context.fillStyle;
    const strokeStyle = context.strokeStyle;
    context.fillStyle = SCROLLBAR_COLOR;
    context.strokeStyle = "rgba(255,255,255,0.8)";
    [scrollBars.horizontal, scrollBars.vertical].forEach((scrollBar) => {
      if (scrollBar) {
        roundRect(
          context,
          scrollBar.x,
          scrollBar.y,
          scrollBar.width,
          scrollBar.height,
          SCROLLBAR_WIDTH / 2,
        );
      }
    });
    context.fillStyle = fillStyle;
    context.strokeStyle = strokeStyle;
  }

  context.scale(1 / scale, 1 / scale);

  return { atLeastOneVisibleElement: visibleElements.length > 0, scrollBars };
}

function isVisibleElement(
  element: ExcalidrawElement,
  viewportWidth: number,
  viewportHeight: number,
  {
    scrollX,
    scrollY,
    zoom,
  }: {
    scrollX: FlooredNumber;
    scrollY: FlooredNumber;
    zoom: number;
  },
) {
  const [x1, y1, x2, y2] = getElementAbsoluteCoords(element);

  // Apply zoom
  const viewportWidthWithZoom = viewportWidth / zoom;
  const viewportHeightWithZoom = viewportHeight / zoom;

  const viewportWidthDiff = viewportWidth - viewportWidthWithZoom;
  const viewportHeightDiff = viewportHeight - viewportHeightWithZoom;

  return (
    x2 + scrollX - viewportWidthDiff / 2 >= 0 &&
    x1 + scrollX - viewportWidthDiff / 2 <= viewportWidthWithZoom &&
    y2 + scrollY - viewportHeightDiff / 2 >= 0 &&
    y1 + scrollY - viewportHeightDiff / 2 <= viewportHeightWithZoom
  );
}

// This should be only called for exporting purposes
export function renderSceneToSvg(
  elements: readonly ExcalidrawElement[],
  rsvg: RoughSVG,
  svgRoot: SVGElement,
  {
    offsetX = 0,
    offsetY = 0,
  }: {
    offsetX?: number;
    offsetY?: number;
  } = {},
) {
  if (!svgRoot) {
    return;
  }
  // render elements
  elements.forEach((element) => {
    if (!element.isDeleted) {
      renderElementToSvg(
        element,
        rsvg,
        svgRoot,
        element.x + offsetX,
        element.y + offsetY,
      );
    }
  });
}
