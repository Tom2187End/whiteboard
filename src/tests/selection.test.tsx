import React from "react";
import ReactDOM from "react-dom";
import { render, fireEvent } from "./test-utils";
import { App } from "../index";
import * as Renderer from "../renderer/renderScene";
import { KEYS } from "../keys";

// Unmount ReactDOM from root
ReactDOM.unmountComponentAtNode(document.getElementById("root")!);

const renderScene = jest.spyOn(Renderer, "renderScene");
beforeEach(() => {
  localStorage.clear();
  renderScene.mockClear();
});

describe("selection element", () => {
  it("create selection element on mouse down", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("selection");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;
    fireEvent.mouseDown(canvas, { clientX: 60, clientY: 100 });

    expect(renderScene).toHaveBeenCalledTimes(1);
    const selectionElement = renderScene.mock.calls[0][1]!;
    expect(selectionElement).not.toBeNull();
    expect(selectionElement.type).toEqual("selection");
    expect([selectionElement.x, selectionElement.y]).toEqual([60, 100]);
    expect([selectionElement.width, selectionElement.height]).toEqual([0, 0]);

    // TODO: There is a memory leak if mouse up is not triggered
    fireEvent.mouseUp(canvas);
  });

  it("resize selection element on mouse move", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("selection");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;
    fireEvent.mouseDown(canvas, { clientX: 60, clientY: 100 });
    fireEvent.mouseMove(canvas, { clientX: 150, clientY: 30 });

    expect(renderScene).toHaveBeenCalledTimes(2);
    const selectionElement = renderScene.mock.calls[1][1]!;
    expect(selectionElement).not.toBeNull();
    expect(selectionElement.type).toEqual("selection");
    expect([selectionElement.x, selectionElement.y]).toEqual([60, 30]);
    expect([selectionElement.width, selectionElement.height]).toEqual([90, 70]);

    // TODO: There is a memory leak if mouse up is not triggered
    fireEvent.mouseUp(canvas);
  });

  it("remove selection element on mouse up", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("selection");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;
    fireEvent.mouseDown(canvas, { clientX: 60, clientY: 100 });
    fireEvent.mouseMove(canvas, { clientX: 150, clientY: 30 });
    fireEvent.mouseUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(3);
    const selectionElement = renderScene.mock.calls[2][1];
    expect(selectionElement).toBeNull();
  });
});

describe("select single element on the scene", () => {
  it("rectangle", () => {
    const { getByToolName, container } = render(<App />);
    const canvas = container.querySelector("canvas")!;
    {
      // create element
      const tool = getByToolName("rectangle");
      fireEvent.click(tool);
      fireEvent.mouseDown(canvas, { clientX: 30, clientY: 20 });
      fireEvent.mouseMove(canvas, { clientX: 60, clientY: 70 });
      fireEvent.mouseUp(canvas);
      fireEvent.keyDown(document, { key: KEYS.ESCAPE });
    }

    const tool = getByToolName("selection");
    fireEvent.click(tool);
    // click on a line on the rectangle
    fireEvent.mouseDown(canvas, { clientX: 45, clientY: 20 });
    fireEvent.mouseUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(7);
    const elements = renderScene.mock.calls[6][0];
    const selectionElement = renderScene.mock.calls[6][1];
    expect(selectionElement).toBeNull();
    expect(elements.length).toEqual(1);
    expect(elements[0].isSelected).toBeTruthy();
  });

  it("diamond", () => {
    const { getByToolName, container } = render(<App />);
    const canvas = container.querySelector("canvas")!;
    {
      // create element
      const tool = getByToolName("diamond");
      fireEvent.click(tool);
      fireEvent.mouseDown(canvas, { clientX: 30, clientY: 20 });
      fireEvent.mouseMove(canvas, { clientX: 60, clientY: 70 });
      fireEvent.mouseUp(canvas);
      fireEvent.keyDown(document, { key: KEYS.ESCAPE });
    }

    const tool = getByToolName("selection");
    fireEvent.click(tool);
    // click on a line on the rectangle
    fireEvent.mouseDown(canvas, { clientX: 45, clientY: 20 });
    fireEvent.mouseUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(7);
    const elements = renderScene.mock.calls[6][0];
    const selectionElement = renderScene.mock.calls[6][1];
    expect(selectionElement).toBeNull();
    expect(elements.length).toEqual(1);
    expect(elements[0].isSelected).toBeTruthy();
  });

  it("ellipse", () => {
    const { getByToolName, container } = render(<App />);
    const canvas = container.querySelector("canvas")!;
    {
      // create element
      const tool = getByToolName("ellipse");
      fireEvent.click(tool);
      fireEvent.mouseDown(canvas, { clientX: 30, clientY: 20 });
      fireEvent.mouseMove(canvas, { clientX: 60, clientY: 70 });
      fireEvent.mouseUp(canvas);
      fireEvent.keyDown(document, { key: KEYS.ESCAPE });
    }

    const tool = getByToolName("selection");
    fireEvent.click(tool);
    // click on a line on the rectangle
    fireEvent.mouseDown(canvas, { clientX: 45, clientY: 20 });
    fireEvent.mouseUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(7);
    const elements = renderScene.mock.calls[6][0];
    const selectionElement = renderScene.mock.calls[6][1];
    expect(selectionElement).toBeNull();
    expect(elements.length).toEqual(1);
    expect(elements[0].isSelected).toBeTruthy();
  });

  it("arrow", () => {
    const { getByToolName, container } = render(<App />);
    const canvas = container.querySelector("canvas")!;
    {
      // create element
      const tool = getByToolName("arrow");
      fireEvent.click(tool);
      fireEvent.mouseDown(canvas, { clientX: 30, clientY: 20 });
      fireEvent.mouseMove(canvas, { clientX: 60, clientY: 70 });
      fireEvent.mouseUp(canvas);
      fireEvent.keyDown(document, { key: KEYS.ESCAPE });
    }

    const tool = getByToolName("selection");
    fireEvent.click(tool);
    // click on a line on the rectangle
    fireEvent.mouseDown(canvas, { clientX: 45, clientY: 20 });
    fireEvent.mouseUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(7);
    const elements = renderScene.mock.calls[6][0];
    const selectionElement = renderScene.mock.calls[6][1];
    expect(selectionElement).toBeNull();
    expect(elements.length).toEqual(1);
    expect(elements[0].isSelected).toBeTruthy();
  });

  it("arrow", () => {
    const { getByToolName, container } = render(<App />);
    const canvas = container.querySelector("canvas")!;
    {
      // create element
      const tool = getByToolName("line");
      fireEvent.click(tool);
      fireEvent.mouseDown(canvas, { clientX: 30, clientY: 20 });
      fireEvent.mouseMove(canvas, { clientX: 60, clientY: 70 });
      fireEvent.mouseUp(canvas);
      fireEvent.keyDown(document, { key: KEYS.ESCAPE });
    }

    const tool = getByToolName("selection");
    fireEvent.click(tool);
    // click on a line on the rectangle
    fireEvent.mouseDown(canvas, { clientX: 45, clientY: 20 });
    fireEvent.mouseUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(7);
    const elements = renderScene.mock.calls[6][0];
    const selectionElement = renderScene.mock.calls[6][1];
    expect(selectionElement).toBeNull();
    expect(elements.length).toEqual(1);
    expect(elements[0].isSelected).toBeTruthy();
  });
});
