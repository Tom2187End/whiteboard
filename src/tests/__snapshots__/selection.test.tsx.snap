// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select single element on the scene arrow 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "lastCommittedPoint": null,
  "opacity": 100,
  "points": Array [
    Array [
      0,
      0,
    ],
    Array [
      30,
      50,
    ],
  ],
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "arrow",
  "version": 3,
  "versionNonce": 449462985,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`select single element on the scene arrow escape 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "lastCommittedPoint": null,
  "opacity": 100,
  "points": Array [
    Array [
      0,
      0,
    ],
    Array [
      30,
      50,
    ],
  ],
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "line",
  "version": 3,
  "versionNonce": 449462985,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`select single element on the scene diamond 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "opacity": 100,
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "diamond",
  "version": 2,
  "versionNonce": 1278240551,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`select single element on the scene ellipse 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "opacity": 100,
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "ellipse",
  "version": 2,
  "versionNonce": 1278240551,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`select single element on the scene rectangle 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "opacity": 100,
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "rectangle",
  "version": 2,
  "versionNonce": 1278240551,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;
